fairseq/tasks/masked_lm.py

```diff
@@ -100,6 +100,11 @@ def add_args(parser):
             help="comma-separated list of dataset splits to apply shortening to, "
             'e.g., "train,valid" (default: all dataset splits)',
         )
+        parser.add_argument(
+            "--lang-code",
+            default=None,
+            help="the lang code. Only used for raw text datasets."
+        )
```

Comment:
That might be good enough, honestly. I can try to make a change and see how it works. The nature of `create_dummy_data` was a bit confusing to me, so I tried to make _something_ work. 😆

Reply:
Maybe to take a step back - the reason this was introduced is because of moving to raw datasets. the issue is that unlike having a filename convention for looking up a file given a split (i.e., appending ".bin" or ".idx"), there's no such convention for raw datasets since the default extension is the bitext extensions. Without this, you'll find the tests fail with "file doesn't exist" errors for dataset `blah/blah/blah/{train,valid}`. If you _only_ specify, e.g., `/path/to/data/train.en` as your `data` arg, 1) I'm not sure dictionary loading will work - maybe it will, 2) I'm not sure how to generically infer the files corresponding to the other splits which I think is currently unhandled

Reply:
I actually just tested and it seems like this doesn't work - I might be doing something wrong, but I tried providing `os.path.join(data_dir, "train.in")` as the argument for positional `data` argument in the tests, but it looks like it's expected that it's a directory.

Reply:
Sorry I'm still a bit lost :) This seems somewhat unrelated to Windows support, since Windows should still support the bin/idx files, right? I do think it'd be good to include some test coverage for raw text datasets, but would prefer to discuss that in a separate PR (and have this one just focus on Windows)

Reply:
Windows does support bin and idx, but the tests cannot prove this. It's not easy to explain, but the shortest story:

Preproces, train, validate, and generate succeed with the temp dir.

The temp dir tries to clean itself up.

Because the task gets cleaned up whenever python wants it to, the bin files haven't been closed and the test fails for that reason. This is due to very weird ownership relationships within the Datasets hierarchy.

Reply:
FWIW, `--dataset-impl cached` is not a drop-in replacement for `--dataset-impl mmap`:

```log
tests/test_binaries.py:1271: in train_legacy_masked_language_model
    train.main(train_args)
fairseq_cli/train.py:71: in main
    task.load_dataset(valid_sub_split, combine=False, epoch=1)
fairseq/tasks/cross_lingual_lm.py:167: in load_dataset
    split=language_split, epoch=epoch
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <fairseq.tasks.cross_lingual_lm.CrossLingualLMTask object at 0x7f82a9529390>
split = 'valid.in', epoch = 1

    def _load_single_lang_dataset(self, split, epoch):
        loaded_datasets = []

        paths = utils.split_paths(self.args.data)
        assert len(paths) > 0
        data_path = paths[(epoch - 1) % len(paths)]

        for k in itertools.count():
            split_k = split + (str(k) if k > 0 else "")
            path = os.path.join(data_path, split_k)

            ds = data_utils.load_indexed_dataset(
                path, self.dictionary, self.args.dataset_impl
            )
            if ds is None:
                if k > 0:
                    break
                else:
                    raise FileNotFoundError(
>                       "Dataset not found: {} ({})".format(split, data_path)
                    )
E                   FileNotFoundError: Dataset not found: valid.in (/var/folders/v8/ls8m50ks3zz0hzq21681r8m40000gn/T/tmpcuwyf6pttest_legacy_mlm)
```